### PR TITLE
flow_ebos: fix the OpenMP functionality

### PR DESCRIPTION
**Pull request by andlaus**
the ThreadManager from ebos was not called which resulted in some havoc when attempting multi-threaded runs.

this PR depends on OPM/ewoms#234.

### CODE REVIEW

**Comment by andlaus** [comment]
jenkins build this ewoms=234 please

File: opm/autodiff/FlowMainEbos.hpp [hunk 1]
```diff
@@ -410,14 +411,27 @@ namespace Opm
...
+            argv.push_back("flow_ebos");
+
+            std::string deckFileParam("--ecl-deck-file-name=");
+            deckFileParam += param_.get<std::string>("deck_filename");
+            argv.push_back(deckFileParam.c_str());
+
+#if defined(_OPENMP)
+            std::string numThreadsParam("--threads-per-process=");
+            int numThreads = 1;
+#pragma omp parallel
```
Comment (akva2) [inline opm/autodiff/FlowMainEbos.hpp]: omp_get_max_threads() ? or maybe omp_get_thread_limit() ?

Comment (andlaus) [reply opm/autodiff/FlowMainEbos.hpp]: Maybe, I just copy and pasted [this](https://github.com/OPM/opm-simulators/blob/master/opm/autodiff/FlowMainEbos.hpp#L172). shall I change both to `omp_get_max_threads()`?

Comment (akva2) [reply opm/autodiff/FlowMainEbos.hpp]: that's the equivalent, ie number of threads used for next parallel section. omp_get_thread_limit() is the amount available on start, ie, that is not affected by omp_set_num_threads() calls.

Comment (andlaus) [reply opm/autodiff/FlowMainEbos.hpp]: thanks for the info. update pushed.

**Comment by andlaus** [comment]
jenkins build this ewoms=234 please

File: opm/autodiff/FlowMainEbos.hpp [hunk 2]
```diff
@@ -167,15 +168,12 @@ namespace Opm
                 int num_threads = std::min(4, num_cores);
                 omp_set_num_threads(num_threads);
             }
-#pragma omp parallel
-            if (omp_get_thread_num() == 0) {
```
Comment (atgeirr) [inline opm/autodiff/FlowMainEbos.hpp]: I think this part tries to make sure only one of the OpenMP threads write the message below, why did you remove that?

File: opm/autodiff/FlowMainEbos.hpp [hunk 3]
```diff
@@ -410,14 +408,25 @@ namespace Opm
...
+            int numThreads = omp_get_max_threads();
+
+            numThreadsParam += std::to_string(numThreads);
+            argv.push_back(numThreadsParam.c_str());
+#endif // defined(_OPENMP)
+
             EbosSimulator::registerParameters();
-            Ewoms::setupParameters_< TypeTag > ( 2, ptr );
+            Ewoms::setupParameters_<TypeTag>(argv.size(), &argv[0]);
+            EbosThreadManager::init();
```
Comment (atgeirr) [inline opm/autodiff/FlowMainEbos.hpp]: I guess this statement should be inside #if defined(_OPENMP)?

File: opm/autodiff/FlowMainEbos.hpp (hunk 2 above)
Comment (andlaus) [reply opm/autodiff/FlowMainEbos.hpp]: the proposed code should be identical and it is quite a bit nicer, IMO. see @akva2's [comments](https://github.com/OPM/opm-simulators/pull/1316#discussion_r148505282)

Comment (akva2) [reply opm/autodiff/FlowMainEbos.hpp]: @atgeirr in the new code we avoid the fork into a parallel section, and thus there is only one thread around.

File: opm/autodiff/FlowMainEbos.hpp (hunk 3 above)
Comment (andlaus) [reply opm/autodiff/FlowMainEbos.hpp]: that's not necessary, `Ewoms::ThreadManager` deals with this sort of stuff internally.

File: opm/autodiff/FlowMainEbos.hpp (hunk 2 above)
Comment (atgeirr) [reply opm/autodiff/FlowMainEbos.hpp]: Ok, I see now that it uses omp_get_max_threads() instead of omp_get_num_threads(), more questions:
 - Is it ok to call omp_get_max_threads() outside the parallel section? (Unlike omp_get_num_threads(), as said in the comment, that now should be removed).
 - Will it give equivalent results (i.e. can the user somehow choose a different number of threads than max)? (If so, why do both exist?)

Comment (akva2) [reply opm/autodiff/FlowMainEbos.hpp]: - omp_get_num_threads() returns *the number of currently active threads*. if you call it inside a  parallel section you get the number of spawned threads, if not you naturally get 1 always.
- opm_get_max_threads() returns the number of threads that will be used for the *next* parallel section.
- omp_set_num_threads() controls the returned value of omp_get_max_threads(), ie the number of threads that will be used in the next parallel section. you can never go above omp_get_thread_limit()

Comment (atgeirr) [reply opm/autodiff/FlowMainEbos.hpp]: Thanks for the thorough explanation, finally got it. I'll merge!
